### ISSUE TITLE
AutoMapper 6.0 can't map nullable types containing null when not built using .NET Native tool chain

### ISSUE THREAD

**Issue by mobilemotion** [AutoMapper/AutoMapper#2016]
Mapping an object that contains a nullable property which contains `null` throws an `AutoMapperMappingException` with the following message:

> Error mapping types.
> 
> Mapping types:
> In -> Out
> Workheld.Core.In -> Workheld.Core.Out
> 
> Type Map configuration:
> In -> Out
> Workheld.Core.In -> Workheld.Core.Out
> 
> Property:
> B

The inner exception says

> Non-static method requires a target.

This happens only when building in _Debug_ mode, meaning when the _Compile with .NET Native tool chain_ compiler option is not checked. When using .NET Native tool chain, the mapping works and produces the expected result. Also, when using AutoMapper 5.1.1, the mapping works.

To reproduce, create a Portable Class Library targeting:

- .NET Framework 4.5
- Windows 8
- APS.NET Core 1.0
- Windows Phone 8.1
- Xamarin.Android
- Xamarin.iOS
- Xamarin.iOS (Classic)
- Xamarin.Mac

(Profile111) that references the AutoMapper 6.0.1 NuGet package, and add the following classes:

    public class In
    {
        public int? A { get; set; }
        public int? B { get; set; }
    }
    public class Out
    {
        public int? A { get; set; }
        public int? B { get; set; }
    }
    public static class TestClass
    {
        public static Out TestMethod()
        {
            Mapper.Initialize(c =>
            {
                c.CreateMap<In, Out>();
            });
            var input = new In
            {
                A = 123
            };
            return Mapper.Map<In, Out>(input);
        }
    }

Then, create a blank UWP app project that references the PCL and calls the `TestMethod`. Run it in _Debug_ and _Release_ mode, and watch the different behavior.

**Comment by jbogard** [comment]
Uggghhhhhhh

**Comment by lbargaoanu** [comment]
I think this is a bug in the .NET Core expression interpreter. I don't see anything wrong with the expression we generate.

**Comment by lbargaoanu** [comment]
Here is a repro without AM.

            var hasValue = typeof(int?).GetRuntimeProperty("HasValue");
            var parameter = Parameter(typeof(int?), "param");
            var planTest = Lambda<Func<int?, bool>>(Property(parameter, hasValue), parameter);
            var compiled = planTest.Compile();
            var result = compiled(null);

**Comment by Mikaka27** [comment]
Hi,
Is there a workaround for AM 6.0 until the issue in the expression interpreter is fixed?
Also I'm asking here because I don't know how to reach reporter of this issue.

Mobilemotion did you report this issue in the net core?

**Comment by lbargaoanu** [comment]
You can take over the nullable mappings.

**Comment by mobilemotion** [comment]
@Mikaka27 how did you try to reach me? If there are any questions that don't belong here, feel free to drop me an e-mail, my address should be listed on my profile page!
To answer your question: I didn't have time to review the the underlying issue in the .NET core yet, so I didn't report it. To be honest, since this was not the first issue I've experienced with AM, I took the chance to remove the reference to AM from my project completely, and use an alternative object mapper by now.

**Comment by jbogard** [comment]
@mobilemotion what other issues did you run into? Just issues with UWP or something else?

**Comment by mobilemotion** [comment]
Yeah, all UWP / PCL related. It's just that every version seems to have some issue that makes it not fully useable for me... 5.1.1 made release builds last forever ([https://github.com/AutoMapper/AutoMapper/issues/1655](url)), 5.2.0 can not be referenced at all from a Profile111 PCL, and now 6.0 fails on debug builds (see above).
By the way, although issue 1655 issue is fixed, .NET native builds still take around 7 minutes, which is fine for release builds (especially compared to the 45 minutes with version 5.1.1), but still too much for everyday debugging, so I can't do without a working debug build, that's why I decided to replace the whole library...

**Comment by jbogard** [comment]
Oh well you should be off Profile111 anyway. Silverlight ?!?!

I wish I could debug a build to see why it's so slow, the fix for 45 min ->
7 was actually very trivial once we saw where the issue was.

**Comment by lbargaoanu** [comment]
I think all these issues are because expressions don't work so well on these other platforms, not really about AM.

**Comment by johndbell** [comment]
@lbargaoanu I'm fairly new to AM, and am seeing this issue in Debug using Xamarin Forms in iOS (android is fine). I'm interested to know, when you suggest "take over the nullable mappings" can you explain how to do that for any nullable type on any object being mapped?  Thanks in advance.

**Comment by mobilemotion** [comment]
Since VS2015 even has a shortcut for creating a Profile111 PCL in the _New Project_ wizard, I would't be too sure everyone is off Profile111 by now - this has nothing to do with Silverlight, it just was the default profile for WinRT & Xamarin for a long time. By the way, Profile7 (which seems to be the profile of choice for UWP & Xamarin) isn't supported by AM 5.2.0 either. Let's hope .NETStandard 2.0 will bring an end to the mess of different PCL platforms when it's out finally...

**Comment by jbogard** [comment]
Oh wait, I should be able to work in Profile111 and Profile7 - those are both netstandard1.1 https://docs.microsoft.com/en-us/dotnet/articles/standard/library it should absolutely support those profiles.

**Comment by mobilemotion** [comment]
@jbogard its not a big issue since it affects only 5.2.0, 5.1.1 and 6.0.x work perfectly, but since you mentioned it I took a quick look behind the scenes:
Seems to be related to [https://github.com/dotnet/corefx/issues/10445](url):

- AM 5.1.1 does specify a dedicated DLL for portable-net45+win8+wpa81 with _no dependencies_ - I guess Profile7 and 111 target this one, that's why it works
- In AM 5.2.0, the portable-net45+win8+wpa81 target is gone, so profile Profile7 and 111 target the _netstandard1.1_ lib, which specifies the _System.Dynamic.Runtime 4.0_ dependency, which is where the .NET core issue mentioned above comes in
- AM 6.0 specifies _System.Dynamic.Runtime 4.3_ as dependency for netstandard1.1 targets, which makes it work again

-> When manually installing _System.Runtime.InteropServices.RuntimeInformation 4.3_ first, AM 5,2.0 can be installed without any problems.

**Comment by jbogard** [comment]
Ah, I think I included that one in 5.1.1 because of a nuget thing.

PCL can go die in a fire

**Comment by lbargaoanu** [comment]
@johndbell You need to bypass the NullableSourceMapper. Probably the easiest way is to insert before it a [simple custom mapper](https://github.com/AutoMapper/AutoMapper/blob/50cd30f67bb057f845899cb43a63f40c3c52fbbc/src/UnitTests/Mappers/CustomMapperTests.cs#L69) that does the job.

**Comment by xplatsolutions** [comment]
@lbargaoanu I still don't get it how to fix this, any example code? I get this error in Xamarin iOS and version 6.0.2 PCL Profile 111. @jbogard

**Comment by lbargaoanu** [comment]
If the test I linked to doesn't give you an idea on how to proceed, maybe you should better wait for Xamarin to fix their bug. I think they just put out a new version.

**Comment by xplatsolutions** [comment]
Unfortunately I can't upgrade the the latest Xamarin due to other dependency problems.

**Comment by lbargaoanu** [comment]
Maybe you're lucky. For other reasons, #2115 changes that code and it no longer calls HasValue. So the problem might go away once #2115 is merged.

**Comment by xplatsolutions** [comment]
Cool, I will look out for an updated nuget version.

**Comment by wdupree** [comment]
Has anyone else been able to get past this without code changes? I have a UWP that isn't using Xamarin at all and still cannot map from one nullable to another - but only when the value actually is null, otherwise it works in 6.0.2; if i revert back to 5.2.0 I have no issues at all. I can even use NullSubstitute and assign it something else (in my case, it's a Double? so setting it to 0) works. I really don't know how to get past this and I hate to stay a version behind just because of this issue. Thanks!

**Comment by lbargaoanu** [comment]
Try the [MyGet](https://github.com/AutoMapper/AutoMapper/wiki/The-MyGet-Build) build.

**Comment by premiaware** [comment]
I have the same problem with my UWP App. i have try to install
Install-Package AutoMapper -Version 6.1.0-ci-01328 -Source https://www.myget.org/F/automapperdev/api/v3/index.json 

but not fix the problem.
Thanks!

**Comment by lbargaoanu** [comment]
A repro would help. Make a [gist](https://gist.github.com/) that we can execute and see fail. Also post the stack trace for the exception you get.

**Comment by wdupree** [comment]
FYI - this issue resolved for me in 6.1.1, but I have a new issue that's related but involves non-null values in a nullable property (#2203). Thanks!
